pySDC/helpers/fieldsIO.py

```diff
@@ -0,0 +1,731 @@
+        If a MPI decomposition is used, field **must be** the local field values.
+        """
+        if not self.MPI_ON:
+            return super().addField(time, field)
+
+        assert self.initialized, "cannot add field to a non initialized FieldsIO"
+
+        field = np.asarray(field)
+        assert field.dtype == self.dtype, f"expected {self.dtype} dtype, got {field.dtype}"
+        assert field.shape == (self.nVar, self.nLocX), f"expected {(self.nVar, self.nLocX)} shape, got {field.shape}"
+
+        offset0 = self.fileSize
+        self.MPI_FILE_OPEN(mode="a")
+        if self.MPI_ROOT:
+            self.MPI_WRITE(np.array(time, dtype=T_DTYPE))
+        offset0 += self.tSize
+
+        for iVar in range(self.nVar):
+            offset = offset0 + (iVar * self.nX + self.iLocX) * self.itemSize
+            self.MPI_WRITE_AT(offset, field[iVar])
+        self.MPI_FILE_CLOSE()
+
+    def readField(self, idx):
+        """
+        Read one field stored in the binary file, corresponding to the given
+        time index, eventually in MPI mode.
+
+        Parameters
+        ----------
+        idx : int
+            Positional index of the field.
+
+        Returns
+        -------
+        t : float
+            Stored time for this field.
+        field : np.ndarray
+            Read (local) fields in a numpy array.
+
+        Note
+        ----
+        If a MPI decomposition is used, it reads and returns the local fields values only.
+        """
+        if not self.MPI_ON:
+            return super().readField(idx)
+        idx = self.check(idx)
+
+        offset0 = self.hSize + idx * (self.fSize + self.tSize)
+        with open(self.fileName, "rb") as f:
+            t = float(np.fromfile(f, dtype=T_DTYPE, count=1, offset=offset0)[0])
+        offset0 += self.tSize
+
+        field = np.empty((self.nVar, self.nLocX), dtype=self.dtype)
+
+        self.MPI_FILE_OPEN(mode="r")
+        for iVar in range(self.nVar):
+            offset = offset0 + (iVar * self.nX + self.iLocX) * self.itemSize
+            self.MPI_READ_AT(offset, field[iVar])
+        self.MPI_FILE_CLOSE()
+
+        return t, field
+
+
+@FieldsIO.register(sID=2)
+class Cart2D(Cart1D):
+    """FieldsIO handler storing a given number of 2D scalar variables"""
+
+    # -------------------------------------------------------------------------
+    # Overriden methods
+    # -------------------------------------------------------------------------
+    def setHeader(self, nVar, gridX, gridY):
+        """
+        Set the descriptive grid structure to be stored in the file header.
+
+        Parameters
+        ----------
+        nVar : int
+            Number of 1D variables stored.
+        gridX : np.1darray
+            The grid coordinates in x direction.
+        gridY : np.1darray
+            The grid coordinates in y direction.
+
+        Note
+        ----
+        When used in MPI decomposition mode, `gridX` and `gridY` **must** be the global grid.
```

Comment:
Is there a way to enforce this by raising an error? This note seems easy to miss.. Also, would be nice to give only the local grid here. Getting the global grid would have been quite difficult for my large GPU Gray-Scott run..

Reply:
This may a bit tricky ... both checking if the grid is global (and not local) and the possibility to give the local grid only requires additional all-to-all communications. And at then end, those are only 1D arrays that each processors can totally generate on their own, since normally each process start with the same problem parameters.

So in the attempt of making stuff simple, I would simply add some dedicated section in the module docstring, and blame any user for not reading it.

Reply:
Can you explain why the fields can be written to disk in parallel, but the grid cannot?

Reply:
Grids are encoded in the header, that is only written by rank 0. Since it's only done once when creating the file, this makes things way easier. Also when writing collectively, each process needs to know the global size of the problem, which is directly determined from the global grid that every process stores.

Reply:
I suppose we can merge as is then and eventually find a find a different solution with distributed writing of the grid once we need it.

Reply:
Did you got that the grids are simply the 1D coordinates in each directions, not the full N-dimensional grid ? Because if not, maybe the naming could be corrected :sweat_smile:

Reply:
I did not get that! Ok, that is all fine then.

Reply:
Then I'll change the naming to `coord[...]`, which make be more explicit ... don't merge it yet !